lean/main/overview.lean

```diff
@@ -0,0 +1,200 @@
+
+Throughout this book you will see references to the elaborator; and in the "Extra: Pretty Printing" chapter you can read on delaborators.
+
+## 3 essential functions and their syntax sugars
+
+Now, when you're reading Lean source code, you will see 11(+?) commands specifying the **parsing**/**elaboration**/**evaluation** process.  Most of them are syntax sugar, and you only need **3 commands** to do metaprogramming in Lean:
+
+- **syntax rule: `syntax (name := xxx) ... : command`**
+
+    For example, `syntax (name := xxx) "#help" "option" (ident)? : command` is a syntax rule that will match (remember the regex analogy) all strings in the form of `"#help option hi.hello"` or just `"#help option"`.   
+    Other widespread syntax categories are `tactic` and `term`, all of these are used in different physical places in your code.
+
+- **macro: `@[macro xxx] def ourMacro : Macro := ...`**
+
+    For example, ``@[macro xxx] def ourMacro : Macro := λ stx => match stx with | _ => `(tactic| pick_goal 2)``. Whenever your Lean code matches the syntax rule with the name `xxx`, this macro will be applied.  
+    If the syntax rule was `syntax (name := xxx) "swap" : tactic`, this macro will syntactically turn `swap` into `pick_goal 2` - and the subsequent elaboration step will be handled by the `pick_goal 2` syntax rule.
+
+- **elab: `@[command_elab xxx] def ourElab : CommandElab := ...`**
+
+    For example, ``@[term_elab our_help] def ourElab : TermElab := λ stx tp => (Expr.app (Expr.const `Nat.add []))``.  
+    Our `elab` function can be of different types - the **CommandElab** and **TermElab** that you saw already, and **Tactic**.  
+    **TermElab** stands for **Syntax → Option Expr → TermElabM Expr**, so the elaboration function is expected to return the **Expr** object.  
+    **CommandElab** stands for **Syntax → CommandElabM Unit**, so it shouldn't return anything.  
+    **Tactic** stands for **Syntax → TacticM Unit**, so it shouldn't return anything either.  
+    This corresponds to out intuitive understanding of terms, commands and tactics in Lean - terms return a particular value upon execution, commands modify the environment or print something out, and tactics modify the proof state.
+
+These `syntax (name := xxx) ... : command`, `@[macro xxx] def ourMacro : Macro := ...` and `@[command_elab xxx] def ourElab : CommandElab := ...` are the 3 essential, low-level commands, and you can get away with them only. Lean standard library and Mathlib use many syntax sugars, however, so memorizing them is well worth the effort. I'm summing them up in the next diagram.
+
+<p align="center">
+<img width="650px" src="https://github.com/arthurpaulino/lean4-metaprogramming-book/assets/7578559/38e9d3fd-af93-4f89-b17b-e56a3b13244a"/>
+</p>
+
+In the image above:
```

Comment:
IMO both image and text are a bit cryptic and I'm not sure what should be the reader take away from this section. If it's to mention that multiple syntactic sugar constructions are provided then probably just mentioning that is as helpful. Other option might be to demonstrate on examples instead of placeholders `~` so the reader can match what visual/logical "shorcut" each syntactic sugar provides

Reply:
I agree, in fact I changed to that image from this one:

![image](https://github.com/arthurpaulino/lean4-metaprogramming-book/assets/7578559/d8a9477d-576e-4937-a3f0-80d59a5da45c)

because I didn't think that one was helpful.

I think in the new image it would indeed be possible to fit the examples, will try that.

Reply:
<img width="877" alt="image" src="https://github.com/arthurpaulino/lean4-metaprogramming-book/assets/7578559/da54dba9-d974-462b-8bfe-031d0b6c774a">

Reply:
Do you think any of these is better than the current version?

Reply:
<img width="907" alt="image" src="https://github.com/arthurpaulino/lean4-metaprogramming-book/assets/7578559/896d56ed-2e29-4033-8d1d-628e3ba706df">

Reply:
I like the last one.
One thing which is not clear: Does `Sugar for syntax + macro` means I don't need to write syntax from first table and it combines both? If so probably worth to put an empty box or some other indication that syntax isn't needed to be specified separately in the first column (maybe arrow) to the left of syntax + macros/syntax + elab

Reply:
With the last picture there is some illusion created that each row is supposed to be used with each other, which is not really true.
I also want to fit `notation`/`prefix`/`infix`/`postfix` here to make it clear that these are [syntax + macro]s.

I think there should be two separate images, one as an overview, and one as implementing #help.

Overview:
![image](https://github.com/arthurpaulino/lean4-metaprogramming-book/assets/7578559/9b83f06c-49c4-4d93-9d42-72e0499ae6c8)

Implementing #help:
![image](https://github.com/arthurpaulino/lean4-metaprogramming-book/assets/7578559/c3a81507-a477-4dd3-96a7-3560367cb52f)

Reply:
> Does Sugar for syntax + macro means I don't need to write syntax from first column and it combines both?

Yes. Tried with the empty box, it got more confusing, and the implication of "you should read row by row" more pronounced.